pytensor/scalar/loop.py

```diff
@@ -0,0 +1,388 @@
+                "If you want to return an output as a lagged input, wrap it in an identity Op."
+            )
+
+    @property
+    def fgraph(self):
+        if hasattr(self, "_fgraph"):
+            return self._fgraph
+
+        fgraph = FunctionGraph(self.inputs, self.outputs)
+        # TODO: We could convert to TensorVariable, optimize graph,
+        # and then convert back to ScalarVariable.
+        # This would introduce rewrites like `log(1 + x) -> log1p`.
+        MergeOptimizer().rewrite(fgraph)
+        self._validate_fgraph(fgraph)
+
+        # Clone identical outputs that have been merged
+        if len(set(fgraph.outputs)) != len(self.outputs):
+            old_outputs = fgraph.outputs
+            new_outputs = []
+            for output in old_outputs:
+                if output not in new_outputs:
+                    new_outputs.append(output)
+                else:
+                    node = output.owner
+                    output_idx = node.outputs.index(output)
+                    new_output = node.clone().outputs[output_idx]
+                    new_outputs.append(new_output)
+            fgraph = FunctionGraph(fgraph.inputs, new_outputs, clone=False)
+
+        self._fgraph = fgraph
+        return self._fgraph
+
+    def clone(self):
+        if self.is_while:
+            *update, until = self.outputs
+        else:
+            update, until = self.outputs, None
+        init = self.inputs[: len(update)]
+        constant = self.inputs[len(update) :]
+        return ScalarLoop(
+            init=init,
+            update=update,
+            constant=constant,
+            until=until,
+            until_condition_failed=self.until_condition_failed,
+            name=self.name,
+        )
+
+    @property
+    def fn(self):
+        raise NotImplementedError
+
+    def make_new_inplace(self, output_types_preference=None, name=None):
+        """
+        This op.__init__ fct don't have the same parameter as other scalar op.
+        This break the insert_inplace_optimizer optimization.
+        This fct allow fix patch this.
+
+        """
+        d = {k: getattr(self, k) for k in self.init_param}
+        out = self.__class__(**d)
+        if name:
+            out.name = name
+        else:
+            name = out.name
+        super(ScalarLoop, out).__init__(output_types_preference, name)
+        return out
+
+    def make_node(self, n_steps, *inputs):
+        assert len(inputs) == self.nin - 1
+
+        n_steps = as_scalar(n_steps)
+        if not n_steps.type.dtype.startswith("int"):
+            raise TypeError(
+                "The first variable of ScalarLoop (n_steps) must be of integer type. "
+                f"Got {n_steps.type.dtype}",
+            )
+
+        if self.inputs_type == tuple([i.type for i in inputs]):
+            return super().make_node(n_steps, *inputs)
+        else:
+            # Make a new op with the right input types.
+            res = rebuild_collect_shared(
```

Comment:
Good question... I have to think about this one

Reply:
I guess I can add a test case with a Composite or another Loop inside and see what happens. If it fails, I will just error out explicitly, otherwise... free functionality!

Reply:
I don't see why we need to clone Ops that have InnerGraphs... so I will leave it to False for now. I added two tests, one with an inner composite and another with an inner loop and it seems to work well.